TestStaticRelays is racy

```
=== RUN   TestStaticRelays
  ==================
  WARNING: DATA RACE
  Read at 0x00c00242f8c0 by goroutine 118:
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).handleNewCandidate()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:285 +0xea
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:139 +0x9e4
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).Start.func1()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:537 +0x90
  
  Previous write at 0x00c00242f8c0 by goroutine 28:
    runtime.mapassign_faststr()
        /opt/hostedtoolcache/go/1.16.15/x64/src/runtime/map_faststr.go:202 +0x0
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).handleNewNode()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:[220](https://github.com/libp2p/go-libp2p/runs/5962063510?check_suite_focus=true#step:9:220) +0x684
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).findNodes.func1()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:180 +0x109
  
  Goroutine 118 (running) created at:
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).Start()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:535 +0x23d
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*AutoRelay).background()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/autorelay.go:103 +0x3d2
    github.com/libp2p/go-libp2p/p2p/host/autorelay.NewAutoRelay.func1()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/autorelay.go:59 +0x71
  
  Goroutine 28 (finished) created at:
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).findNodes()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:178 +0x5da
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background.func1()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:101 +0x90
  ==================
      testing.go:1102: race detected during execution of test
  --- FAIL: TestStaticRelays (2.50s)

--- a/p2p/host/autorelay/relay_finder.go
+++ b/p2p/host/autorelay/relay_finder.go
@@ -281,12 +281,16 @@ func (rf *relayFinder) tryNode(ctx context.Context, pi peer.AddrInfo) (supportsR
 
 func (rf *relayFinder) handleNewCandidate(ctx context.Context) {
 	rf.relayMx.Lock()
-	defer rf.relayMx.Unlock()
-	if len(rf.candidates) == 0 {
+	numRelays := len(rf.relays)
+	rf.relayMx.Unlock()
+	// We're already connected to our desired number of relays. Nothing to do here.
+	if numRelays == rf.conf.desiredRelays {
 		return
 	}
-	// We're already connected to our desired number of relays. Nothing to do here.
-	if len(rf.relays) == rf.conf.desiredRelays {
+
+	rf.candidateMx.Lock()
+	defer rf.candidateMx.Unlock()
+	if len(rf.candidates) == 0 {
 		return
 	}
 
@@ -468,12 +472,10 @@ func (rf *relayFinder) usingRelay(p peer.ID) bool {
 // selectCandidates returns an ordered slice of relay candidates.
 // Callers should attempt to obtain reservations with the candidates in this order.
 func (rf *relayFinder) selectCandidates() []*candidate {
-	rf.candidateMx.Lock()
 	var candidates []*candidate
 	for _, cand := range rf.candidates {
 		candidates = append(candidates, cand)
 	}
-	rf.candidateMx.Unlock()
 
 	// TODO: better relay selection strategy; this just selects random relays,
 	// but we should probably use ping latency as the selection metric
@@ -541,7 +543,9 @@ func (rf *relayFinder) Start() error {
 
 func (rf *relayFinder) Stop() error {
 	log.Debug("stopping relay finder")
-	rf.ctxCancel()
+	if rf.ctxCancel != nil {
+		rf.ctxCancel()
+	}
 	rf.refCount.Wait()
 	rf.ctxCancel = nil
 	return nil

